firedrake/preconditioners/pmg.py

```diff
@@ -0,0 +1,300 @@
+from functools import partial
+from itertools import chain
+
+from ufl import MixedElement, FiniteElement, VectorElement, TensorElement, replace
+from ufl.algorithms import map_integrands
+
+from pyop2 import op2
+
+from firedrake.petsc import PETSc
+from firedrake.preconditioners.base import PCBase
+from firedrake.preconditioners.low_order import ArgumentReplacer
+from firedrake.dmhooks import attach_hooks, get_appctx, push_appctx, pop_appctx
+from firedrake.dmhooks import add_hook, get_parent, push_parent, pop_parent
+from firedrake.dmhooks import get_function_space, set_function_space
+from firedrake.solving_utils import _SNESContext
+import firedrake
+
+
+class PMGPC(PCBase):
```

Comment:
Does this subsume the functionality offered by `P1PC`? If yes, can you replace that with the obvious from here?

Reply:
Yes, it does, at the cost of an API change. I have done this. The default PETSc settings are slightly different because e.g. PCMG does pre- and post-smoothing by default. The iteration counts are now beautifully regular (constant in p for a given mesh).